File: backend/models/article.py
from app import db
from datetime import datetime


class Article(db.Model):
    __tablename__ = 'articles'

    id = db.Column(db.Integer, primary_key=True)
    url = db.Column(db.Text, unique=True, nullable=False)
    title = db.Column(db.Text)
    excerpt = db.Column(db.Text)
    author = db.Column(db.Text)
    published_date = db.Column(db.Text)
    saved_date = db.Column(db.DateTime, default=datetime.utcnow)
    tags = db.Column(db.Text, default='[]')
    is_read = db.Column(db.Boolean, default=False)
    thumbnail_url = db.Column(db.Text)

    def to_dict(self):
        return {
            'id': self.id,
            'url': self.url,
            'title': self.title,
            'excerpt': self.excerpt,
            'author': self.author,
            'published_date': self.published_date,
            'saved_date': self.saved_date.isoformat() if self.saved_date else None,
            'tags': self.tags,
            'is_read': self.is_read,
            'thumbnail_url': self.thumbnail_url,
        }

File: backend/models/Article.js
const mongoose = require('mongoose');

// Define the Article schema
const articleSchema = new mongoose.Schema({
  url: { type: String, required: true },
  title: { type: String, default: 'Untitled' },
  excerpt: { type: String, default: '' },
  author: { type: String, default: '' },
  thumbnail_url: { type: String, default: '/logo.png' },
  tags: { type: [String], default: [] },
  // Categories: array of ObjectId references to Category collection
  categories: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Category' }],
  is_read: { type: Boolean, default: false },
  status: { type: String, default: 'pending', enum: ['pending', 'completed', 'failed'] }, // New field
  error_message: { type: String, default: '' }, // New field
  user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null, index: true }, // User reference for multi-user support
  created_at: { type: Date, default: Date.now },
});

// Compound unique index: same URL can exist for different users
articleSchema.index({ url: 1, user_id: 1 }, { unique: true });

// Additional performance indexes
articleSchema.index({ user_id: 1, created_at: -1 }); // For sorting by date
articleSchema.index({ user_id: 1, status: 1 }); // For filtering by status
articleSchema.index({ user_id: 1, is_read: 1 }); // For filtering read/unread
articleSchema.index({ categories: 1 }); // For category filtering

// Create the Article model
const Article = mongoose.model('Article', articleSchema);

// Export the model
module.exports = Article;
